features/26-mfa/api.md

```diff
@@ -0,0 +1,1003 @@
+# Overview
+
+The following factors are supported:
+
+- Time-Based One-Time Password [(TOTP)](https://tools.ietf.org/html/rfc6238)
+- Out-of-band
+  - SMS
+  - Email
+- Recovery code
+- Bearer token
+
+In the future, the below factors will be supported:
+
+- [WebAuthn with CTAP2 or FIDO U2F](https://fidoalliance.org/specifications/)
+
+The developer can configure the enforcement of MFA.
+
+The session contains how the user was authenticated.
+
+# Configuration
+
+```yaml
+mfa:
+  # If enabled is false, the MFA API is disabled. THe user cannot register authenticator.
+  # Default is false.
+  enabled: false
+  # Valid values are 'off', 'optional' or 'required'.
+  # Default is 'optional'.
+  #
+  # 'off' means MFA is never required during authentication.
+  #
+  # 'optional' means MFA is required during authentication when the user has at least 1 activated authenticator. This is like how other major online services such as Google offer MFA.
+  #
+  # 'required' means MFA is always required during authentication.
+  enforcement: 'optional'
+  # The total maximum number of authenticator the user can register.
+  # Valid values is [0, 15].
+  # Default is 1.
+  maximum: 1
+  totp:
+    # The maximum number of TOTP authenticator the user can register.
+    # Valid values is [0, 5].
+    # Default is 0.
+    maximum: 0
```

Comment:
should we make a default value that once `enabled` is `true`, the default value could be used for normal case?

Reply:
I want the developer to choose which authenticator their application supports. The UI of TOTP authenticator and Out-of-Band authenticator are different. If `enable` is `true`, the default value of TOTP authenticator and Out-of-Band authenticator both become `1`, the application is assumed to support both authenticators. I do not want to make such assumption.

If either one kind of authenticator becomes `1`, then the developer have to explicitly set it to `0` to disable it. That also means we flavor that kind of authenticator.